jerry-core/ecma/builtin-objects/ecma-builtins.c

```diff
@@ -949,8 +950,17 @@ ecma_builtin_dispatch_routine (ecma_builtin_id_t builtin_object_id, /**< built-i
                                ecma_length_t arguments_list_len) /**< length of arguments' list */
 {
   JERRY_ASSERT (builtin_object_id < ECMA_BUILTIN_ID__COUNT);
+
+  const ecma_value_t normalized_argumets_list_p[3] =
+  {
+    arguments_list_len >= 1 ? arguments_list_p[0] : ECMA_VALUE_UNDEFINED,
+    arguments_list_len >= 2 ? arguments_list_p[1] : ECMA_VALUE_UNDEFINED,
+    arguments_list_len >= 3 ? arguments_list_p[2] : ECMA_VALUE_UNDEFINED
+  };
+
```

Comment:
Wouldn't it be possible to use
```c
if (arguments_list_len <= 2)
{
  ecma_value_t padded_argumets_list_p[3] = { ECMA_VALUE_UNDEFINED, ECMA_VALUE_UNDEFINED, ECMA_VALUE_UNDEFINED };

  switch (arguments_list_len)
  {
    case 2: padded_argumets_list_p[1] = arguments_list_p[1];
    case 1: padded_argumets_list_p[0] = arguments_list_p[0];
    default: JERRY_ASSERT (arguments_list_len == 0);
  }

  return ecma_builtin_routines[builtin_object_id] (builtin_routine_id,
                                                   this_arg_value,
                                                   padded_arguments_list_p,
                                                   arguments_list_len);
}

return ecma_builtin_routines[builtin_object_id] (builtin_routine_id,
                                                 this_arg_value,
                                                 arguments_list_p,
                                                 arguments_list_len);
```
This way there would be no need for an additional parameter to `ecma_builtin_dispatch_routine_t` functions but they could always be sure that `arguments_list` had at least 3 elements, even if `arguments_number` was less than 3.

I think this is quite close to what @zherczeg suggested.

Reply:
Ohh! Thank you for the example code, thanks to it I see what you were thinking for.